src/verifast1.ml

```diff
@@ -2709,45 +2709,60 @@ module VerifyProgram1(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
+      | Some(targenv) -> {fl; fgh; ft = replace_type dummy_loc targenv ft; fvis; fbinding; ffinal; finit; fvalue}
+    in
+    let replace_super_targs tparams (super, superTargs) =
+      let tpenv = match zip tparams targs with
+        None -> [] (* Static field, or no targs. *)
+      | Some (tpenv) -> tpenv 
+      in
+      match tpenv with
+        [] -> (super, superTargs)
+      | tpenv -> (super, List.map (replace_type dummy_loc tpenv) superTargs) 
+    in
     match try_assoc cn classmap1 with
-      Some (_, _, _, _, fds, _, (super, superTargs), _, itfs, _, _, _) ->
+      Some (_, _, _, _, fds, _, super, tparams, itfs, _, _, _) ->
       begin match try_assoc fn fds with
         None when cn = "java.lang.Object" -> None
-      | Some f -> Some (f, cn)
+      | Some f -> Some (replace_type_param f tparams, cn)
       | None ->
-      match lookup_class_field super fn with
+      match lookup_class_field (replace_super_targs tparams super) fn with
         Some _ as result -> result
       | None ->
-      head_flatmap_option (fun cn -> lookup_class_field cn fn) (List.map fst itfs)
+      head_flatmap_option (fun itf -> lookup_class_field (replace_super_targs tparams itf) fn) itfs
       end
     | None -> 
     match try_assoc cn classmap0 with
-      Some {cfds; csuper=(csuper, _); cinterfs} ->
+      Some {cfds; csuper; cinterfs; ctpenv} ->
       begin match try_assoc fn cfds with
         None when cn = "java.lang.Object" -> None
-      | Some f -> Some (f, cn)
+      | Some f -> Some (replace_type_param f ctpenv, cn)
       | None ->
-      match lookup_class_field csuper fn with
+      match lookup_class_field (replace_super_targs ctpenv csuper) fn with
         Some _ as result -> result
       | None ->
-      head_flatmap_option (fun cn -> lookup_class_field cn fn) (List.map fst cinterfs)
+      head_flatmap_option (fun itf -> lookup_class_field (replace_super_targs ctpenv itf) fn) cinterfs
       end
     | None ->
     match try_assoc cn interfmap1 with
-      Some (_, fds, _, _, supers, _, _, _) ->
+      Some (_, fds, _, _, supers, _, _, tparams) ->
       begin match try_assoc fn fds with
-        Some f -> Some (f, cn)
+        Some f -> Some (replace_type_param f tparams, cn)
       | None ->
-      head_flatmap_option (fun cn -> lookup_class_field cn fn) (List.map fst supers)
+      head_flatmap_option (fun itf -> lookup_class_field (replace_super_targs tparams itf) fn) supers
       end
     | None ->
     match try_assoc cn interfmap0 with
-      Some (InterfaceInfo (_, fds, _, _, supers, _)) ->
+      Some (InterfaceInfo (_, fds, _, _, supers, tparams)) ->
       begin match try_assoc fn fds with
-        Some f -> Some (f, cn)
+        Some f -> Some (replace_type_param f tparams, cn)
       | None ->
-      head_flatmap_option (fun cn -> lookup_class_field cn fn) (List.map fst supers)
+      head_flatmap_option (fun itf -> lookup_class_field (replace_super_targs tparams itf) fn) supers
       end

```

Comment:
Dit is vier keer dezelfde logica. Kan je dit schrijven met minder codeduplicatie?